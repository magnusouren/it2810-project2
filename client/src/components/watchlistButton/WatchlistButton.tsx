import { useMutation } from '@apollo/client';
import { BookmarkAdd, BookmarkRemove } from '@mui/icons-material';
import { ToggleButton } from '@mui/material';
import { FC, memo, useState } from 'react';

import { ADD_MOVIE_TO_WATCHLIST, REMOVE_MOVIE_FROM_WATCHLIST } from '../../graphql/queries';
import { Movie, User } from '../../types';
import styles from './WatchlistButton.module.scss';

interface WatchlistButtonProps {
  movie: Movie;
  style?: 'big' | 'small';
  user: User;
}

const Button: FC<WatchlistButtonProps> = ({ movie, style = 'small', user }) => {
  const movieId = parseInt(movie._id.toString(), 10);
  const baseVariables = { movieId: movieId, userId: user.id };

  const [isInWatchlist, setIsInWatchlist] = useState<boolean>(movie.isInWatchlist ?? false);

  const [removeMovieFromWatchlist, { error: removeError }] = useMutation(REMOVE_MOVIE_FROM_WATCHLIST, {
    variables: baseVariables,
    update(cache) {
      cache.modify({
        fields: {
          getMovieById(existingMovie) {
            if (existingMovie._id === movieId) {
              return { ...existingMovie, isInWatchlist: false }; // Update the cache to reflect the change in the UI
            }
          },
        },
      });
    },
  });

  const [addMovieToWatchlist, { error: addError }] = useMutation(ADD_MOVIE_TO_WATCHLIST, {
    variables: baseVariables,
    update(cache) {
      cache.modify({
        fields: {
          getMovieById(existingMovie) {
            if (existingMovie._id === movieId) {
              return { ...existingMovie, isInWatchlist: true }; // Update the cache to reflect the change in the UI
            }
          },
        },
      });
    },
  });

  if (removeError) {
    return <div>{removeError.message}</div>;
  }

  if (addError) {
    return <div>{addError.message}</div>;
  }

  const handleClick = async () => {
    if (isInWatchlist) {
      const response = await removeMovieFromWatchlist();
      if (response?.data?.removeMovieFromWatchlist) {
        setIsInWatchlist(false);
      }
    } else {
      const response = await addMovieToWatchlist();
      if (response?.data?.addMovieToWatchlist) {
        setIsInWatchlist(true);
      }
    }
  };

  return (
    <div className={style === 'big' ? styles.bigContainer : styles.smallContainer} data-testid='watchlist-button'>
      <ToggleButton
        value='check'
        selected={isInWatchlist}
        onChange={handleClick}
        color='primary'
        size='small'
        aria-label={isInWatchlist ? 'Remove movie from watchlist' : 'Add movie to watchlist'}
        data-testid='watchlist-toggle-button'
        title={isInWatchlist ? 'Remove movie from watchlist' : 'Add movie to watchlist'}
      >
        {isInWatchlist ? <BookmarkRemove className={styles.remove} /> : <BookmarkAdd className={styles.add} />}
        {style === 'big' && (
          <span className={styles.text}>{isInWatchlist ? 'Remove from watchlist' : 'Add to watchlist'}</span>
        )}
      </ToggleButton>
    </div>
  );
};

const WatchlistButton = memo(Button);
export default WatchlistButton;
