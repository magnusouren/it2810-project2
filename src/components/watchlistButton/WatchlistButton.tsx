import { BookmarkAdd, BookmarkRemove } from '@mui/icons-material';
import { ToggleButton } from '@mui/material';
import { FC, memo, useState } from 'react';

import { Movie } from '../../types';
import styles from './WatchlistButton.module.scss';

interface WatchlistButtonProps {
  movie: Movie;
  existInWatchlist: (movie: Movie) => boolean;
  toggleMovieInWatchlist: (movie: Movie) => void;
  style?: 'big' | 'small';
}

const Button: FC<WatchlistButtonProps> = ({ movie, existInWatchlist, toggleMovieInWatchlist, style = 'small' }) => {
  const [selected, setSelected] = useState(() => existInWatchlist(movie));

  return (
    <div className={style == 'big' ? styles.bigContainer : styles.smallContainer}>
      <ToggleButton
        value='check'
        selected={selected}
        color='primary'
        size='small'
        onChange={() => {
          setSelected(!selected);
          toggleMovieInWatchlist(movie);
        }}
      >
        {selected ? <BookmarkRemove className={styles.remove} /> : <BookmarkAdd className={styles.add} />}
      </ToggleButton>
    </div>
  );
};

const WatchlistButton = memo(Button);

export default WatchlistButton;
