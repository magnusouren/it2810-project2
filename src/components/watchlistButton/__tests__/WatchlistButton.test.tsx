import { render } from '@testing-library/react';

import WatchlistButton from '../WatchlistButton';

describe('WatchlistButton', () => {
  const existInWatchlist = () => true;
  const toggleMovieInWatchlist = () => null;

  it('Should match snapshot', () => {
    const { container } = render(
      <WatchlistButton
        movieId={123}
        existInWatchlist={existInWatchlist}
        toggleMovieInWatchlist={toggleMovieInWatchlist}
      />,
    );
    expect(container).toMatchSnapshot();
  });
});
