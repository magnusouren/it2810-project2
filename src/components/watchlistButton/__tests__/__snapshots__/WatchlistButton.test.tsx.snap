// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`WatchlistButton > Should match snapshot 1`] = `
<div>
  <button
    aria-pressed="true"
    class="MuiButtonBase-root MuiToggleButton-root Mui-selected MuiToggleButton-sizeMedium MuiToggleButton-standard css-ueukts-MuiButtonBase-root-MuiToggleButton-root"
    tabindex="0"
    type="button"
    value="check"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
      data-testid="CheckIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M9 16.17 4.83 12l-1.42 1.41L9 19 21 7l-1.41-1.41z"
      />
    </svg>
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
</div>
`;
